### PR TITLE
Implements: [C++] constructor pragma improvement (fix #21921)

### LINKED ISSUE #21921: [C++] constructor pragma improvement
### Summary

This feature requests aims to improve C++ interoperability by allowing to declare and implement C++ constructor from Nim by reusing the existing `constructor` pragma. 

### Description

When a proc is marked with `constructor` and `importcpp` it will behave as always, affecting calls so it prevents double initialization.  
When using without `importcpp`, on the declaration side of things it will behave the way `virtual` does, this is by generating a constructor with the same signature of the function by default and accepting a syntax to alter its params if necessary (i.e. '&&', 'const' etc). If a constructoris declared with parameters, it will generate a default constructor without params so it's legal to init the type.  

Consider,

`proc makeFoo(a : int) : Foo {.constructor.}` 

It will create a new constructor for `Foo` (notice Foo cant be a ptr here, and we really should not allow ptr with `importcpp` neither as it does nothing).

The body wont have a typical `result` field but a `this` which will be `ptr Foo`.  

Edit: Support to call the base class `constructor:"'0('1 #1) : CppClass(#1) "`
https://discord.com/channels/371759389889003530/768367394547957761/1111796672775860266

### Alternatives

_No response_

### Examples

_No response_

### Backwards Compatibility

_No response_

### Links

_No response_

### PR DESCRIPTION
**Pull request by jmgomez**
Implements: [C++] constructor pragma improvement (fix #21921)

### CODE REVIEW

File: compiler/ccgtypes.nim [hunk 1]
```diff
@@ -536,7 +541,13 @@ proc genVirtualProcParams(m: BModule; t: PType, rettype, params: var string,
...
   multiFormat(params, @['\'', '#'], [types, names])
   if params == "()":
-    params = "(void)"
+    if types.len == 0:
+      params = "(void)"
+    else:
+      if not isCtor:
+        names.delete(0)
+        types.delete(0)
+      params = "(" & zip(types, names).mapIt(it[0] & " " & it[1]).join(", ") & ")"
```
Comment (juancarlospaco) [inline compiler/ccgtypes.nim]: This line with `zip` + `mapIt` + `join` feels kinda golfed for the compiler core,
can it be like a for loop that appends to a string or something like that?. 🤔

File: compiler/ccgtypes.nim [hunk 2]
```diff
@@ -687,17 +697,25 @@ proc genRecordFieldsAux(m: BModule; n: PNode,
...
-      genVirtualProcHeader(m, prc, header, false, true)
+      if sfConstructor in prc.flags:
+        isCtorGen = true
+        if prc.typ.n.len == 1:
+          isDefaultCtorGen = true
+      genMemberProcHeader(m, prc, header, false, true)
       result.add "\t" & header & ";\n"
+    if isCtorGen and not isDefaultCtorGen:
+      var ch : IntSet
+      result.add "\t" & getTypeDescAux(m, typ, ch, dkOther) & "() = default;\n"
```
Comment (juancarlospaco) [inline compiler/ccgtypes.nim]: I think you have to use `"$n"` instead of `"\n"` here because not all OS use the same line endings or something like that.

Comment (Araq) [reply compiler/ccgtypes.nim]: No, there is a soft ($n) vs hard ($N) newline abstraction. Hard new lines ($N) are always kept, soft newlines can be skipped if activated for better debug info generation.

Comment (juancarlospaco) [reply compiler/ccgtypes.nim]: Yes, thanks, what I mean is not to hardcode `"\n"`.

Comment (jmgomez) [reply compiler/ccgtypes.nim]: @Araq do you see anything else I should change or you didn't go over it yet?
Some light on why testament is failing on win would be appreciated. That's all this PR is missing as far I can tell

Comment (juancarlospaco) [reply compiler/ccgtypes.nim]: Ive seen random flaky tests on Windows, I wonder if we are occasionally hitting the max file open limit on Windows?.

Comment (jmgomez) [reply compiler/ccgtypes.nim]: In local it outputs this before failing. It fails with a bunch of tests. 
```
warning: LF will be replaced by CRLF in C:\Users\juama\AppData\Local\Temp\diffStrings_b_FUEzC3aD.
The file will have its original line endings in your working directory
diff --git "a/C:\\Users\\juama\\AppData\\Local\\Temp\\diffStrings_a_YiVgiPft" "b/C:\\Users\\juama\\AppData\\Local\\Temp\\diffStrings_b_FUEzC3aD"
index 031aaaad3..88f2ffa0c 100644
--- "a/C:\\Users\\juama\\AppData\\Local\\Temp\\diffStrings_a_YiVgiPft"
+++ "b/C:\\Users\\juama\\AppData\\Local\\Temp\\diffStrings_b_FUEzC3aD"
@@ -1 +1,3 @@
-exitcode: 0
\ No newline at end of file
+exitcode: 1
+
+Output:
```

**Comment by Araq** [comment]
FAIL: tests/cpp/tconstructor.nim cpp 

Your very own test fails. Test it locally via 

nim c -r testament/testament run tests/cpp/tconstructor.nim

**Comment by jmgomez** [comment]
What Im trying to say is that testament is failing for me running it locally (latest devel included) for pretty much any test that I have tried :/

Edit: Found something. For some reason when compiling with `cpp` in my local machine, `echo` doesnt print anything. By passing `cc:clang` to testament they work. Hopefully  with that will be able to repro it.

File: lib/std/private/dragonbox.nim [hunk 3]
```diff
@@ -34,6 +34,7 @@ const
 ##
 ##  It is a simplified version of the reference implementation found here:
 ##  https://github.com/jk-jeon/dragonbox
+##  https://github.com/jk-jeon/dragonbox
```
Comment (Araq) [inline lib/std/private/dragonbox.nim]:
```suggestion
```

File: compiler/ccgtypes.nim [hunk 4]
```diff
@@ -1106,46 +1123,63 @@ proc isReloadable(m: BModule; prc: PSym): bool =
...
-  var name, params, rettype: string
+  var name, params, rettype, superCall: string
   var isFnConst, isOverride: bool
-  parseVFunctionDecl(prc.constraint.strVal, name, params, rettype, isFnConst, isOverride)
-  genVirtualProcParams(m, prc.typ, rettype, params, check, true, false) 
+  parseVFunctionDecl(prc.constraint.strVal, name, params, rettype, superCall, isFnConst, isOverride, isCtor)
+  genMemberProcParams(m, prc, superCall, rettype, params, check, true, false) 
   var fnConst, override: string
+  if isCtor:
+      name = typDesc
```
Comment (Araq) [inline compiler/ccgtypes.nim]:
```suggestion
    name = typDesc
```

**Pull request merged**